### ISSUE TITLE
[rt.openssl.org #4271] Enhancement Request: Support TCP Fast Open

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#2472]
Migrated from [rt.openssl.org#4271](https://rt.openssl.org/Ticket/Display.html?id=4271) (status was 'open')

Requestors:
* sara@sinodun.com

Attachments:
* [signature.asc](https://rt.openssl.org/Ticket/Attachment/62429/38462/signature.asc)

From sara@sinodun.com on 2016-01-25 18:24:55:
```
Hi,

I would like to request that support be added to OpenSSL to enable client applications to make use use of TCP Fast Open (https://tools.ietf.org/html/rfc7413 <https://tools.ietf.org/html/rfc7413>) when initiating the TLS handshake on Linux (TCP Fast Open is available in Linux kernel > 4.1). 

This was discussed in detail on the OpenSSL Users list:
https://mta.openssl.org/pipermail/openssl-users/2016-January/002835.html <https://mta.openssl.org/pipermail/openssl-users/2016-January/002835.html>


Regards

Sara.

-------------------------
Sara Dickinson

http://sinodun.com <http://sinodun.com/>

Sinodun Internet Technologies Ltd.
Magdalen Centre
The Oxford Science Park
Oxford
OX4 4GA
U.K.

```

From kurt@roeckx.be on 2016-01-25 18:42:02:
```
On Mon, Jan 25, 2016 at 06:24:55PM +0000, Sara Dickinson via RT wrote:
> Hi,
> 
> I would like to request that support be added to OpenSSL to enable client applications to make use use of TCP Fast Open (https://tools.ietf.org/html/rfc7413 <https://tools.ietf.org/html/rfc7413>) when initiating the TLS handshake on Linux (TCP Fast Open is available in Linux kernel > 4.1). 

I've seen that request, and I have tought about it.  I'm just
wondering if that comes with security consequences, like replay
attacks.  Specially in combination with what they're doing with
TLS 1.3.

The API clearly doesn't support anything like that currently.


Kurt

```

From sara@sinodun.com on 2016-01-26 14:17:57:
```
> On 25 Jan 2016, at 18:42, Kurt Roeckx via RT <rt@openssl.org> wrote:
> 
> On Mon, Jan 25, 2016 at 06:24:55PM +0000, Sara Dickinson via RT wrote:
>> Hi,
>> 
>> I would like to request that support be added to OpenSSL to enable client applications to make use use of TCP Fast Open (https://tools.ietf.org/html/rfc7413 <https://tools.ietf.org/html/rfc7413>) when initiating the TLS handshake on Linux (TCP Fast Open is available in Linux kernel > 4.1). 
> 
> I've seen that request, and I have tought about it.  I'm just
> wondering if that comes with security consequences, like replay
> attacks.  

Not that I am aware of. Iâve seen the question of security vulnerabilities of TFO raised a few times but never with any concrete examples, just speculation. Iâve also observed Chromium using TFO with TLS.  

I can take the question to the TCPM and/or TLS WG if that is helpful?

Sara. 


```

From kurt@roeckx.be on 2016-01-26 21:05:40:
```
On Tue, Jan 26, 2016 at 02:17:57PM +0000, Sara Dickinson via RT wrote:
> 
> > On 25 Jan 2016, at 18:42, Kurt Roeckx via RT <rt@openssl.org> wrote:
> > 
> > On Mon, Jan 25, 2016 at 06:24:55PM +0000, Sara Dickinson via RT wrote:
> >> Hi,
> >> 
> >> I would like to request that support be added to OpenSSL to enable client applications to make use use of TCP Fast Open (https://tools.ietf.org/html/rfc7413 <https://tools.ietf.org/html/rfc7413>) when initiating the TLS handshake on Linux (TCP Fast Open is available in Linux kernel > 4.1). 
> > 
> > I've seen that request, and I have tought about it.  I'm just
> > wondering if that comes with security consequences, like replay
> > attacks.  
> 
> Not that I am aware of. I've seen the question of security vulnerabilities of TFO raised a few times but never with any concrete examples, just speculation. I've also observed Chromium using TFO with TLS.  

It's just that I've seen this speculation at other places, and
didn't understand it.

> I can take the question to the TCPM and/or TLS WG if that is helpful?

I don't think it's needed.


Kurt

```

From rsalz@akamai.com on 2016-01-26 21:38:03:
```
TFO is interesting because it lets UDP-style attacks happen at the TCP level.  Normally you can't do a TCP attack unless you have a valid client IP address.

Imagine connecting once and then sending the syncookie to the botnet.

This might be outside the scope of things OpenSSL cares about and I know recent Linux kernels have some mitigation capabilities.  Also note that the server side should just work with no changes, it's on a TFO client that needs API changes.

```

From dkg@fifthhorseman.net on 2016-01-27 04:52:00:
```
On Tue 2016-01-26 16:37:58 -0500, Salz, Rich wrote:
> TFO is interesting because it lets UDP-style attacks happen at the TCP
> level.  Normally you can't do a TCP attack unless you have a valid
> client IP address.
>
> Imagine connecting once and then sending the syncookie to the botnet.

This suggests that you have on-path capabilities between each of the
reflectors and the victim, right?

If you have on-path capabilities, couldn't you do a similar attack
against a live TCP session?  learn (or create) the sequence number of a
TCP session between each of the reflectors and the target, and
distribute them to the botnet?  Then each member of the botnet sends out
a TCP packet (sequence numbers augmented in some coordinated fashion) to
the reflector that triggers an ACK (and even worse, a data flow) from
the reflector to the victim.

I've never done this, so maybe i've missed some mitigating detail, but
it seems like the same risk with or without TFO.

   --dkg

```

From rsalz@akamai.com on 2016-01-27 14:19:32:
```
> This suggests that you have on-path capabilities between each of the
> reflectors and the victim, right?

I don't think so:  you need the first attacker to get the cookie, then you spread it out.
 
> If you have on-path capabilities, couldn't you do a similar attack against a live
> TCP session?

Different because there you are interrupting a session, whereas with TFO you're starting a new connection and pushing data to the receiving app on a "new" connection.


```

From alessandro@ghedini.me on 2016-02-08 23:48:25:
```
On Mon, Jan 25, 2016 at 06:24:55pm +0000, Sara Dickinson via RT wrote:
> Hi,
> 
> I would like to request that support be added to OpenSSL to enable client applications to make use use of TCP Fast Open (https://tools.ietf.org/html/rfc7413 <https://tools.ietf.org/html/rfc7413>) when initiating the TLS handshake on Linux (TCP Fast Open is available in Linux kernel > 4.1). 
> 
> This was discussed in detail on the OpenSSL Users list:
> https://mta.openssl.org/pipermail/openssl-users/2016-January/002835.html <https://mta.openssl.org/pipermail/openssl-users/2016-January/002835.html>

I took a stab at implementing TFO support for OpenSSL on Linux and OS X at:
https://github.com/ghedo/openssl/commits/fast_open

This only works for the BIO_s_socket() BIO, but could probably be adapted to
BIO_s_connect() as well if needed.

However I'm not particularly happy with the implementation (it's fairly ugly),
and it would probably be easier to implement this on the application side by
overriding the "write" method of whatever BIO is used, instead of trying to
make OpenSSL do it directly.

Cheers
```

**Comment by tmshort** [comment]
TFO + 0-RTT TLSv1.3 is getting some notice... any additional thoughts on this?

**Comment by tmshort** [comment]
@ghedo what happened to your TFO changes?

**Comment by tmshort** [comment]
Part of the original thread:
https://mta.openssl.org/pipermail/openssl-users/2016-January/002835.html

**Comment by ghedo** [comment]
I can't find the branch anywhere, so I'd say the changes are forever lost. Though it wasn't particularly hard to implement, basically just involves changing `BIO_s_connect` and `BIO_s_accept` to use whatever TFO API is provided by the system (probably easier now since Linux provides a better client API).

Of course this assumes that the application actually uses those BIOs, so might not be that useful.

**Comment by tmshort** [comment]
Fixed by #8962, closing.
